testonly/hammer/contents.go

```diff
@@ -0,0 +1,200 @@
+func (m *mapContents) pickKey(prng rand.Source) []byte {
+	if m.empty() {
+		panic("internal error: can't pick a key, map data is empty!")
+	}
+
+	choice := intN(prng, len(m.data))
+	// Need sorted keys for reproduceability.
+	keys := make([]mapKey, 0)
+	for k := range m.data {
+		keys = append(keys, k)
+	}
+	sort.SliceStable(keys, func(i, j int) bool {
+		return bytes.Compare(keys[i][:], keys[j][:]) == -1
+	})
+	for _, k := range keys {
+		if choice == 0 {
+			return k[:]
+		}
+		choice--
+	}
+	panic("internal error: inconsistent count of map contents")
+}
+
+// checkContents compares information returned from the Map against a local copy
+// of the map's contents.
+func (m *mapContents) checkContents(leafInclusions []*trillian.MapLeafInclusion) error {
+	for _, inc := range leafInclusions {
+		leaf := inc.Leaf
+		var key mapKey
+		copy(key[:], leaf.Index)
+		value, ok := m.data[key]
+		if ok {
+			if string(leaf.LeafValue) != value {
+				return fmt.Errorf("got leaf[%v].LeafValue=%q, want %q", key, leaf.LeafValue, value)
+			}
+			if want := "extra-" + value; string(leaf.ExtraData) != want {
+				return fmt.Errorf("got leaf[%v].ExtraData=%q, want %q", key, leaf.ExtraData, want)
+			}
+		} else {
+			if len(leaf.LeafValue) > 0 {
+				return fmt.Errorf("got leaf[%v].LeafValue=%q, want not-present", key, leaf.LeafValue)
+			}
+		}
+	}
+	return nil
+}
+
+// updatedWith returns a new mapContents object that has been updated to include the
+// given leaves and revision.  A nil receiver object is allowed.
+func (m *mapContents) updatedWith(rev uint64, leaves []*trillian.MapLeaf) *mapContents {
+	// Start from previous map contents
+	result := mapContents{rev: int64(rev), data: make(map[mapKey]string)}
+	if m != nil {
+		for k, v := range m.data {
+			result.data[k] = v
+		}
+	}
+	// Update with given leaves
+	for _, leaf := range leaves {
+		var k mapKey
+		copy(k[:], leaf.Index)
+		if leaf.LeafValue != nil {
+			result.data[k] = string(leaf.LeafValue)
+		} else {
+			delete(result.data, k)
+		}
+	}
+
+	return &result
+}
+
+// How many copies of map contents to hold on to.
+const copyCount = 10
+
+type previousMapContents struct {
+	mu sync.RWMutex
+
+	// contents holds copies of the map at different revisions,
+	// from later to earlier (so [0] is the most recent).
+	contents [copyCount]*mapContents
+}
+
+func (p *previousMapContents) empty() bool {
+	return p.lastCopy() == nil
+}
+
+// prevCopy returns the specified copy of the map's contents.
+func (p *previousMapContents) prevCopy(which int) *mapContents {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	return p.contents[which]
+}
+
+// lastCopy returns the most recent copy of the map's contents.
+func (p *previousMapContents) lastCopy() *mapContents {
+	return p.prevCopy(0)
+}
+
+// pickCopy returns a previous copy of the map's contents, returning
+// nil if there are no local copies.
+func (p *previousMapContents) pickCopy(prng rand.Source) *mapContents {
```

Comment:
What's the thinking behind passing `rand.Source` everywhere and not an instance of `rand.Rand` made with `rand.New(prng)`?